code/game/gamemodes/nuclear/pinpointer.dm

```diff
@@ -21,277 +24,117 @@
-				if("DNA")
-					var/DNAstring = input("Input DNA string to search for." , "Please Enter String." , "")
-					if(!DNAstring)
-						return
-					for(var/mob/living/carbon/C in mob_list)
-						if(!C.dna)
-							continue
-						if(C.dna.unique_enzymes == DNAstring)
-							target = C
-							break
-
-			return attack_self()
-
-
-///////////////////////
-//nuke op pinpointers//
-///////////////////////
-
-
-/obj/item/weapon/pinpointer/nukeop
-	var/mode = 0	//Mode 0 locates disk, mode 1 locates the shuttle
-	var/obj/docking_port/mobile/home
-
-
-/obj/item/weapon/pinpointer/nukeop/attack_self(mob/user)
-	if(!active)
-		active = 1
-		var/mode_text = "Authentication Disk Locator mode"
-		if(!mode)
-			workdisk()
-		else
-			mode_text = "Shuttle Locator mode"
-			worklocation()
-		user << "<span class='notice'>You activate the pinpointer([mode_text]).</span>"
-	else
-		active = 0
-		icon_state = "pinoff"
-		user << "<span class='notice'>You deactivate the pinpointer.</span>"
-
-
-/obj/item/weapon/pinpointer/nukeop/workdisk()
-	if(!active) return
-	if(mode)		//Check in case the mode changes while operating
-		worklocation()
-		return
-	if(bomb_set)	//If the bomb is set, lead to the shuttle
-		mode = 1	//Ensures worklocation() continues to work
-		worklocation()
-		playsound(loc, 'sound/machines/twobeep.ogg', 50, 1)	//Plays a beep
-		visible_message("Shuttle Locator mode actived.")			//Lets the mob holding it know that the mode has changed
-		return		//Get outta here
-	scandisk()
-	if(!the_disk)
-		icon_state = "pinonnull"
+	scan_for_target()
+	point_to_target()
+	my_god_jc_a_bomb()
+
+/obj/item/weapon/pinpointer/proc/scan_for_target() //Looks for whatever it's tracking
+	if(target)
+		if(isliving(target))
+			var/mob/living/L = target
+			if(L.stat == DEAD)
+				target = null
 		return
-	setDir(get_dir(src, the_disk))
-	switch(get_dist(src, the_disk))
-		if(0)
-			icon_state = "pinondirect"
-		if(1 to 8)
-			icon_state = "pinonclose"
-		if(9 to 16)
-			icon_state = "pinonmedium"
-		if(16 to INFINITY)
-			icon_state = "pinonfar"
-
-	spawn(5) .()
-
-
-/obj/item/weapon/pinpointer/nukeop/proc/worklocation()
+	switch(mode)
+		if("nuclear") //We track the nuke disk, for protection or more nefarious purposes
+			var/obj/item/weapon/disk/nuclear/N = locate()
+			target = N
+		if("malf_ai") //We track the AI, who wants to blow us all to smithereens
+			for(var/mob/living/silicon/ai/A in ai_list)
+				if(A.nuking)
+					target = A
+			for(var/obj/machinery/power/apc/A in apcs_list)
+				if(A.malfhack && A.occupier)
+					target = A
+		if("infiltrator") //We track the Syndicate infiltrator, so we can find our way back
+			target = SSshuttle.getShuttle("syndicate")
+		if("operative")
+			var/list/possible_targets = list()
+			for(var/datum/mind/M in ticker.mode.syndicates)
+				if(M.current && M.current.stat != DEAD && M.current.client)
+					possible_targets |= M.current
+			if(possible_targets.len)
+				target = pick(possible_targets)

```

Comment:
Trying my best to do this as I type this comment out, but Byond doesn't like associative lists and is difficult. If you have any advice, I'd love to hear it.